File: sdk/go/flashq/pool.go
package flashq

import (
	"bufio"
	"context"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"io"
	"net"
	"sync"
	"sync/atomic"
	"time"

	"github.com/vmihailenco/msgpack/v5"
)

var (
	bufferPool = sync.Pool{New: func() interface{} { return make([]byte, 0, 4096) }}
	mapPool    = sync.Pool{New: func() interface{} { return make(map[string]interface{}, 8) }}
)

// PooledConnection is a single connection in the pool with health tracking.
type PooledConnection struct {
	conn      net.Conn
	reader    *bufio.Reader
	writer    *bufio.Writer
	mu        sync.Mutex
	healthy   atomic.Bool
	lastUse   atomic.Int64
	lastCheck atomic.Int64
	failures  atomic.Int32
}

func (pc *PooledConnection) isHealthy() bool { return pc.healthy.Load() }
func (pc *PooledConnection) markHealthy()    { pc.healthy.Store(true); pc.failures.Store(0) }
func (pc *PooledConnection) markUnhealthy()  { pc.healthy.Store(false); pc.failures.Add(1) }

// ConnectionPool manages multiple connections with auto-reconnect.
type ConnectionPool struct {
	opts        ClientOptions
	connections []*PooledConnection
	size        int
	index       atomic.Uint64
	mu          sync.RWMutex
	connected   atomic.Bool
	ctx         context.Context
	cancel      context.CancelFunc
	wg          sync.WaitGroup

	// Metrics
	totalReconnects atomic.Int64
	totalFailures   atomic.Int64
}

// NewConnectionPool creates a pool with specified size.
func NewConnectionPool(opts ClientOptions, size int) *ConnectionPool {
	if size < 1 {
		size = 4
	}
	ctx, cancel := context.WithCancel(context.Background())
	return &ConnectionPool{opts: opts, connections: make([]*PooledConnection, size), size: size, ctx: ctx, cancel: cancel}
}

// Connect establishes all connections in the pool.
func (p *ConnectionPool) Connect(ctx context.Context) error {
	p.mu.Lock()
	defer p.mu.Unlock()

	if p.connected.Load() {
		return nil
	}

	for i := 0; i < p.size; i++ {
		pc, err := p.createConnection(ctx)
		if err != nil {
			p.closeAllLocked()
			return err
		}
		p.connections[i] = pc
	}

	p.connected.Store(true)

	// Start health checker
	p.wg.Add(1)
	go p.healthChecker()

	return nil
}

func (p *ConnectionPool) createConnection(ctx context.Context) (*PooledConnection, error) {
	addr := fmt.Sprintf("%s:%d", p.opts.Host, p.opts.Port)
	dialer := net.Dialer{Timeout: p.opts.ConnectTimeout}

	conn, err := dialer.DialContext(ctx, "tcp", addr)
	if err != nil {
		return nil, NewConnectionError("failed to connect", err)
	}

	if tc, ok := conn.(*net.TCPConn); ok {
		tc.SetNoDelay(true)
		tc.SetKeepAlive(true)
		tc.SetKeepAlivePeriod(30 * time.Second)
		tc.SetWriteBuffer(65536)
		tc.SetReadBuffer(65536)
	}

	pc := &PooledConnection{
		conn:   conn,
		reader: bufio.NewReaderSize(conn, 65536),
		writer: bufio.NewWriterSize(conn, 65536),
	}
	pc.healthy.Store(true)
	pc.lastUse.Store(time.Now().UnixNano())

	if p.opts.Token != "" {
		if err := p.authConn(pc); err != nil {
			conn.Close()
			return nil, err
		}
	}

	return pc, nil
}

func (p *ConnectionPool) authConn(pc *PooledConnection) error {
	resp, err := p.sendOnConn(pc, map[string]interface{}{"cmd": "AUTH", "token": p.opts.Token}, p.opts.Timeout)
	if err != nil {
		return err
	}
	if success, ok := resp["success"].(bool); !ok || !success {
		return NewAuthenticationError("authentication failed")
	}
	return nil
}

// healthChecker periodically checks and reconnects unhealthy connections.
func (p *ConnectionPool) healthChecker() {
	defer p.wg.Done()
	ticker := time.NewTicker(5 * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-p.ctx.Done():
			return
		case <-ticker.C:
			p.checkAndReconnect()
		}
	}
}

func (p *ConnectionPool) checkAndReconnect() {
	p.mu.Lock()
	defer p.mu.Unlock()

	for i, pc := range p.connections {
		if pc == nil || !pc.isHealthy() || pc.failures.Load() > 0 {
			// Try to reconnect
			if pc != nil && pc.conn != nil {
				pc.conn.Close()
			}
			newPc, err := p.createConnection(p.ctx)
			if err == nil {
				p.connections[i] = newPc
				p.totalReconnects.Add(1)
			}
		}
	}
}

// reconnectConn attempts to reconnect a specific connection with backoff.
func (p *ConnectionPool) reconnectConn(idx int) {
	p.mu.Lock()
	defer p.mu.Unlock()

	pc := p.connections[idx]
	if pc != nil && pc.conn != nil {
		pc.conn.Close()
	}

	// Exponential backoff
	maxAttempts := p.opts.MaxReconnectAttempts
	if maxAttempts == 0 {
		maxAttempts = 10
	}
	delay := p.opts.ReconnectDelay
	if delay == 0 {
		delay = time.Second
	}
	maxDelay := p.opts.MaxReconnectDelay
	if maxDelay == 0 {
		maxDelay = 30 * time.Second
	}

	for attempt := 0; attempt < maxAttempts; attempt++ {
		select {
		case <-p.ctx.Done():
			return
		default:
		}

		newPc, err := p.createConnection(p.ctx)
		if err == nil {
			p.connections[idx] = newPc
			p.totalReconnects.Add(1)
			return
		}

		p.totalFailures.Add(1)
		time.Sleep(delay)
		delay *= 2
		if delay > maxDelay {
			delay = maxDelay
		}
	}
}

// Close closes all connections.
func (p *ConnectionPool) Close() error {
	p.cancel()
	p.wg.Wait()

	p.mu.Lock()
	defer p.mu.Unlock()
	p.connected.Store(false)
	return p.closeAllLocked()
}

func (p *ConnectionPool) closeAllLocked() error {
	var lastErr error
	for i, pc := range p.connections {
		if pc != nil && pc.conn != nil {
			if err := pc.conn.Close(); err != nil {
				lastErr = err
			}
			p.connections[i] = nil
		}
	}
	return lastErr
}

// IsConnected returns pool status.
func (p *ConnectionPool) IsConnected() bool { return p.connected.Load() }

// Stats returns pool statistics.
func (p *ConnectionPool) Stats() (reconnects, failures int64, healthy int) {
	p.mu.RLock()
	defer p.mu.RUnlock()
	for _, pc := range p.connections {
		if pc != nil && pc.isHealthy() {
			healthy++
		}
	}
	return p.totalReconnects.Load(), p.totalFailures.Load(), healthy
}

// getConn gets next healthy connection (round-robin with fallback).
func (p *ConnectionPool) getConn() (*PooledConnection, int) {
	start := p.index.Add(1)
	for i := 0; i < p.size; i++ {
		idx := int((start + uint64(i)) % uint64(p.size))
		pc := p.connections[idx]
		if pc != nil && pc.isHealthy() {
			return pc, idx
		}
	}
	// Return any connection as fallback
	idx := int(start % uint64(p.size))
	return p.connections[idx], idx
}

// Send sends command on next available connection with auto-retry.
func (p *ConnectionPool) Send(cmd map[string]interface{}, timeout time.Duration) (map[string]interface{}, error) {
	if !p.connected.Load() {
		return nil, NewConnectionError("not connected", nil)
	}

	pc, idx := p.getConn()
	if pc == nil {
		return nil, NewConnectionError("no available connections", nil)
	}

	resp, err := p.sendOnConn(pc, cmd, timeout)
	if err != nil {
		pc.markUnhealthy()
		// Try reconnect in background
		go p.reconnectConn(idx)
		// Retry on another connection
		pc2, _ := p.getConn()
		if pc2 != nil && pc2 != pc {
			return p.sendOnConn(pc2, cmd, timeout)
		}
		return nil, err
	}

	pc.markHealthy()
	return resp, nil
}

func (p *ConnectionPool) sendOnConn(pc *PooledConnection, cmd map[string]interface{}, timeout time.Duration) (map[string]interface{}, error) {
	pc.mu.Lock()
	defer pc.mu.Unlock()

	if pc.conn == nil {
		return nil, NewConnectionError("connection closed", nil)
	}

	if timeout > 0 {
		pc.conn.SetDeadline(time.Now().Add(timeout))
		defer pc.conn.SetDeadline(time.Time{})
	}

	if p.opts.UseBinary {
		return p.sendBinaryOnConn(pc, cmd)
	}
	return p.sendJSONOnConn(pc, cmd)
}

// sendJSONOnConn sends command using JSON newline-delimited protocol.
func (p *ConnectionPool) sendJSONOnConn(pc *PooledConnection, cmd map[string]interface{}) (map[string]interface{}, error) {
	data, err := json.Marshal(cmd)
	if err != nil {
		return nil, NewValidationError(fmt.Sprintf("marshal error: %v", err))
	}

	buf := bufferPool.Get().([]byte)[:0]
	buf = append(buf, data...)
	buf = append(buf, '\n')
	_, err = pc.writer.Write(buf)
	bufferPool.Put(buf[:0])

	if err != nil {
		return nil, NewConnectionError("write error", err)
	}
	if err := pc.writer.Flush(); err != nil {
		return nil, NewConnectionError("flush error", err)
	}

	line, err := pc.reader.ReadBytes('\n')
	if err != nil {
		return nil, NewConnectionError("read error", err)
	}

	resp := mapPool.Get().(map[string]interface{})
	for k := range resp {
		delete(resp, k)
	}
	if err := json.Unmarshal(line, &resp); err != nil {
		mapPool.Put(resp)
		return nil, NewServerError(fmt.Sprintf("parse error: %v", err), 0)
	}
	if errMsg, ok := resp["error"].(string); ok && errMsg != "" {
		mapPool.Put(resp)
		return nil, NewServerError(errMsg, 0)
	}

	pc.lastUse.Store(time.Now().UnixNano())
	return resp, nil
}

// sendBinaryOnConn sends command using MessagePack binary protocol.
// Protocol: 4-byte big-endian length prefix + msgpack data
func (p *ConnectionPool) sendBinaryOnConn(pc *PooledConnection, cmd map[string]interface{}) (map[string]interface{}, error) {
	// Encode command with msgpack
	data, err := msgpack.Marshal(cmd)
	if err != nil {
		return nil, NewValidationError(fmt.Sprintf("msgpack marshal error: %v", err))
	}

	// Write 4-byte length prefix (big-endian) + data
	lenBuf := make([]byte, 4)
	binary.BigEndian.PutUint32(lenBuf, uint32(len(data)))

	if _, err := pc.writer.Write(lenBuf); err != nil {
		return nil, NewConnectionError("write length error", err)
	}
	if _, err := pc.writer.Write(data); err != nil {
		return nil, NewConnectionError("write data error", err)
	}
	if err := pc.writer.Flush(); err != nil {
		return nil, NewConnectionError("flush error", err)
	}

	// Read 4-byte length prefix
	respLenBuf := make([]byte, 4)
	if _, err := io.ReadFull(pc.reader, respLenBuf); err != nil {
		return nil, NewConnectionError("read length error", err)
	}
	respLen := binary.BigEndian.Uint32(respLenBuf)

	// Sanity check: max 100MB response
	if respLen > 100*1024*1024 {
		return nil, NewServerError(fmt.Sprintf("response too large: %d bytes", respLen), 0)
	}

	// Read msgpack data
	respData := make([]byte, respLen)
	if _, err := io.ReadFull(pc.reader, respData); err != nil {
		return nil, NewConnectionError("read data error", err)
	}

	// Decode response
	resp := mapPool.Get().(map[string]interface{})
	for k := range resp {
		delete(resp, k)
	}
	if err := msgpack.Unmarshal(respData, &resp); err != nil {
		mapPool.Put(resp)
		return nil, NewServerError(fmt.Sprintf("msgpack parse error: %v", err), 0)
	}
	if errMsg, ok := resp["error"].(string); ok && errMsg != "" {
		mapPool.Put(resp)
		return nil, NewServerError(errMsg, 0)
	}

	pc.lastUse.Store(time.Now().UnixNano())
	return resp, nil
}

File: sdk/python/flashq/client/pool.py
"""
flashQ Connection Pool

High-performance connection pool with health checks and auto-reconnect.
"""

import asyncio
import random
from dataclasses import dataclass, field
from typing import Any

from .connection import FlashQConnection
from ..types import ClientOptions
from ..errors import ConnectionError
from ..utils.logger import Logger


@dataclass
class PooledConnection:
    """A connection in the pool with health tracking."""

    connection: FlashQConnection
    healthy: bool = True
    failures: int = 0
    last_use: float = 0.0

    def mark_healthy(self) -> None:
        self.healthy = True
        self.failures = 0

    def mark_unhealthy(self) -> None:
        self.healthy = False
        self.failures += 1


@dataclass
class PoolStats:
    """Connection pool statistics."""

    total_connections: int = 0
    healthy_connections: int = 0
    total_reconnects: int = 0
    total_failures: int = 0


class ConnectionPool:
    """
    Connection pool for flashQ with health checks.

    Features:
    - Multiple connections with round-robin selection
    - Background health checking
    - Auto-reconnect with exponential backoff
    - Request retry on connection failure

    Example:
        ```python
        pool = ConnectionPool(options, pool_size=4)
        await pool.connect()

        response = await pool.send({"cmd": "PUSH", ...})

        await pool.close()
        ```
    """

    DEFAULT_POOL_SIZE = 4
    HEALTH_CHECK_INTERVAL = 5.0  # seconds

    def __init__(
        self,
        options: ClientOptions | None = None,
        pool_size: int | None = None,
    ):
        self._options = options or ClientOptions()
        self._pool_size = pool_size or self._options.pool_size or self.DEFAULT_POOL_SIZE
        self._connections: list[PooledConnection] = []
        self._index = 0
        self._lock = asyncio.Lock()
        self._connected = False
        self._closing = False
        self._health_task: asyncio.Task[None] | None = None

        # Metrics
        self._total_reconnects = 0
        self._total_failures = 0

        self._logger = Logger(
            level=self._options.log_level,
            prefix="Pool",
        )

    @property
    def connected(self) -> bool:
        return self._connected

    @property
    def pool_size(self) -> int:
        return self._pool_size

    async def connect(self) -> None:
        """Connect all connections in the pool."""
        if self._connected:
            return

        async with self._lock:
            if self._connected:
                return

            # Create and connect all connections
            for i in range(self._pool_size):
                conn = FlashQConnection(self._options)
                try:
                    await conn.connect()
                    pc = PooledConnection(
                        connection=conn,
                        healthy=True,
                        last_use=asyncio.get_event_loop().time(),
                    )
                    self._connections.append(pc)
                except Exception as e:
                    # Close already created connections
                    for pc in self._connections:
                        await pc.connection.close()
                    self._connections.clear()
                    raise ConnectionError(f"Failed to create pool: {e}")

            self._connected = True
            self._logger.debug(
                "Pool connected",
                size=self._pool_size,
            )

            # Start health checker
            self._health_task = asyncio.create_task(self._health_checker())

    async def close(self) -> None:
        """Close all connections in the pool."""
        self._closing = True
        self._connected = False

        # Stop health checker
        if self._health_task:
            self._health_task.cancel()
            try:
                await self._health_task
            except asyncio.CancelledError:
                pass
            self._health_task = None

        # Close all connections
        async with self._lock:
            for pc in self._connections:
                try:
                    await pc.connection.close()
                except Exception:
                    pass
            self._connections.clear()

        self._logger.debug("Pool closed")

    def stats(self) -> PoolStats:
        """Get pool statistics."""
        healthy = sum(1 for pc in self._connections if pc.healthy)
        return PoolStats(
            total_connections=len(self._connections),
            healthy_connections=healthy,
            total_reconnects=self._total_reconnects,
            total_failures=self._total_failures,
        )

    async def send(
        self,
        command: dict[str, Any],
        timeout: int | None = None,
    ) -> dict[str, Any]:
        """
        Send command on next available connection.

        Automatically retries on different connection if first fails.
        """
        if not self._connected:
            raise ConnectionError("Pool not connected")

        # Get next healthy connection (round-robin)
        pc, idx = self._get_connection()
        if pc is None:
            raise ConnectionError("No available connections")

        try:
            response = await pc.connection._send(command, timeout)
            pc.mark_healthy()
            pc.last_use = asyncio.get_event_loop().time()
            return response

        except Exception as e:
            pc.mark_unhealthy()
            self._total_failures += 1

            # Try another connection
            pc2, _ = self._get_connection(exclude_idx=idx)
            if pc2 is not None and pc2 != pc:
                try:
                    response = await pc2.connection._send(command, timeout)
                    pc2.mark_healthy()
                    return response
                except Exception:
                    pc2.mark_unhealthy()

            raise ConnectionError(f"All connections failed: {e}")

    def _get_connection(
        self,
        exclude_idx: int | None = None,
    ) -> tuple[PooledConnection | None, int]:
        """Get next healthy connection using round-robin."""
        if not self._connections:
            return None, -1

        # Try to find a healthy connection
        for i in range(len(self._connections)):
            idx = (self._index + i) % len(self._connections)
            if exclude_idx is not None and idx == exclude_idx:
                continue

            pc = self._connections[idx]
            if pc.healthy and pc.connection.connected:
                self._index = (idx + 1) % len(self._connections)
                return pc, idx

        # Fallback to any connection
        self._index = (self._index + 1) % len(self._connections)
        idx = self._index
        if exclude_idx is not None and idx == exclude_idx:
            idx = (idx + 1) % len(self._connections)
        return self._connections[idx], idx

    async def _health_checker(self) -> None:
        """Background task to check and reconnect unhealthy connections."""
        while not self._closing:
            try:
                await asyncio.sleep(self.HEALTH_CHECK_INTERVAL)

                if self._closing:
                    break

                await self._check_and_reconnect()

            except asyncio.CancelledError:
                break
            except Exception as e:
                self._logger.warn("Health check error", error=str(e))

    async def _check_and_reconnect(self) -> None:
        """Check connections and reconnect unhealthy ones."""
        for i, pc in enumerate(self._connections):
            if self._closing:
                break

            if not pc.healthy or not pc.connection.connected or pc.failures > 0:
                await self._reconnect_connection(i)

    async def _reconnect_connection(self, idx: int) -> None:
        """Reconnect a specific connection with backoff."""
        if idx >= len(self._connections):
            return

        pc = self._connections[idx]

        # Close old connection
        try:
            await pc.connection.close()
        except Exception:
            pass

        # Exponential backoff
        max_attempts = self._options.max_reconnect_attempts or 10
        base_delay = (self._options.reconnect_delay or 1000) / 1000  # to seconds
        max_delay = (self._options.max_reconnect_delay or 30000) / 1000

        for attempt in range(max_attempts):
            if self._closing:
                return

            delay = min(base_delay * (2 ** attempt), max_delay)
            jitter = random.uniform(0, 0.1 * delay)

            await asyncio.sleep(delay + jitter)

            try:
                new_conn = FlashQConnection(self._options)
                await new_conn.connect()

                self._connections[idx] = PooledConnection(
                    connection=new_conn,
                    healthy=True,
                    last_use=asyncio.get_event_loop().time(),
                )
                self._total_reconnects += 1

                self._logger.info(
                    "Reconnected",
                    connection=idx,
                    attempt=attempt + 1,
                )
                return

            except Exception as e:
                self._logger.warn(
                    "Reconnection failed",
                    connection=idx,
                    attempt=attempt + 1,
                    error=str(e),
                )

        self._logger.error(
            "Max reconnection attempts reached",
            connection=idx,
        )
